File: problems/31-bellman-ford/002-negative-cycle-variants/java/Solution.java
import java.util.*;
public class Solution {
    static final long INF = Long.MAX_VALUE / 4;
    public long[] shortestPaths(int n, int[][] edges, int source) {
        long[] d = new long[n]; Arrays.fill(d, INF); d[source] = 0;
        for (int i=1;i<n;i++) { boolean changed=false; for (int[] e:edges) if (d[e[0]]<INF && d[e[1]]>d[e[0]]+e[2]) { d[e[1]]=d[e[0]]+e[2]; changed=true; } if (!changed) break; }
        for (int[] e:edges) if (d[e[0]]<INF && d[e[1]]>d[e[0]]+e[2]) throw new IllegalArgumentException("Reachable negative cycle");
        return d;
    }
}

File: problems/31-bellman-ford/002-negative-cycle-variants/python/solution.py
class Solution:
    def shortest_paths(self, n, edges, source):
        inf = 10**30
        dist = [inf] * n
        dist[source] = 0
        for _ in range(n - 1):
            changed = False
            for u, v, w in edges:
                if dist[u] < inf and dist[v] > dist[u] + w:
                    dist[v] = dist[u] + w
                    changed = True
            if not changed:
                break
        for u, v, w in edges:
            if dist[u] < inf and dist[v] > dist[u] + w:
                raise ValueError('Reachable negative cycle')
        return dist
